core/chains/evm/chain.go

```diff
@@ -0,0 +1,353 @@
+	l := opts.Logger.With("chainID", chainID.String())
+	cfg := evmconfig.NewChainScopedConfig(opts.ORM, l, opts.Config, dbchain)
+	if cfg.EVMDisabled() {
+		return nil, errors.Errorf("cannot create new chain with ID %s, EVM is disabled", dbchain.ID.String())
+	}
+	if !dbchain.Enabled {
+		return nil, errors.Errorf("cannot create new chain with ID %s, the chain is disabled", dbchain.ID.String())
+	}
+	if err := cfg.Validate(); err != nil {
+		return nil, errors.Wrapf(err, "cannot create new chain with ID %s, config validation failed", dbchain.ID.String())
+	}
+	db := opts.DB
+	serviceLogLevels, err := l.GetServiceLogLevels()
+	if err != nil {
+		return nil, err
+	}
+	var client eth.Client
+	if cfg.EthereumDisabled() {
+		client = &eth.NullClient{CID: chainID}
+	} else if opts.GenEthClient == nil {
+		var err2 error
+		client, err2 = newEthClientFromChain(l, dbchain)
+		if err2 != nil {
+			return nil, errors.Wrapf(err2, "failed to instantiate eth client for chain with ID %s", dbchain.ID.String())
+		}
+	} else {
+		client = opts.GenEthClient(dbchain)
+	}
+
+	headBroadcaster := headtracker.NewHeadBroadcaster(l)
+	var headTracker httypes.Tracker
+	if cfg.EthereumDisabled() {
+		headTracker = &headtracker.NullTracker{}
+	} else if opts.GenHeadTracker == nil {
+		headTrackerLogger, err2 := l.InitServiceLevelLogger(logger.HeadTracker, serviceLogLevels[logger.HeadTracker])
+		if err2 != nil {
+			return nil, errors.Wrapf(err2, "failed to instantiate head tracker for chain with ID %s", dbchain.ID.String())
+		}
+		orm := headtracker.NewORM(db, *chainID)
+		headTracker = headtracker.NewHeadTracker(headTrackerLogger, client, cfg, orm, headBroadcaster)
+	} else {
+		headTracker = opts.GenHeadTracker(dbchain)
+	}
+
+	var txm bulletprooftxmanager.TxManager
+	if cfg.EthereumDisabled() {
+		txm = &bulletprooftxmanager.NullTxManager{ErrMsg: fmt.Sprintf("Ethereum is disabled for chain %d", chainID)}
+	} else if opts.GenTxManager == nil {
+		txm = bulletprooftxmanager.NewBulletproofTxManager(db, client, cfg, opts.KeyStore, opts.AdvisoryLocker, opts.EventBroadcaster, l)
+	} else {
+		txm = opts.GenTxManager(dbchain)
+	}
+
+	headBroadcaster.Subscribe(txm)
+
+	// Highest seen head height is used as part of the start of LogBroadcaster backfill range
+	highestSeenHead, err := headTracker.HighestSeenHeadFromDB()
+	if err != nil {
+		return nil, err
+	}
+
+	var balanceMonitor services.BalanceMonitor
+	if !cfg.EthereumDisabled() && cfg.BalanceMonitorEnabled() {
+		balanceMonitor = services.NewBalanceMonitor(db, client, opts.KeyStore, l)
+		headBroadcaster.Subscribe(balanceMonitor)
+	}
+
+	var logBroadcaster log.Broadcaster
+	if cfg.EthereumDisabled() {
+		logBroadcaster = &log.NullBroadcaster{ErrMsg: fmt.Sprintf("Ethereum is disabled for chain %d", chainID)}
+	} else if opts.GenLogBroadcaster == nil {
+		logBroadcaster = log.NewBroadcaster(log.NewORM(db, *chainID), client, cfg, l, highestSeenHead)
+	} else {
+		logBroadcaster = opts.GenLogBroadcaster(dbchain)
+	}
```

Comment:
i wonder if usage of these Gen.... methods would be a bit simpler if we always defined them outside of this method? Perhaps have type like ProductionChainDependencies that will contain the logic for the '} else if opts.GenXYZ == nil {' branches, and similarly e.g. EthereumDisabledChainDependencies.

Reply:
Hmm, I think the logic would get a bit difficult to follow if we did that. This way, it's easy to go to one place and see where everything is being instantiated.

Reply:
true, it might - it's quite debatable, I'm ok with either